[TG-2892] Add builtin function class for string_of_int

romainbrenguier opened the pull request:
This makes the analysis of programs with int to string conversion whose result does not affect the control flow of the program more efficient. Tested on my computer, the `checkDet` test runs in 0.15s instead of 3s and the `checkNonDet` in 0.5s instead of 8s.

romainbrenguier committed:
Rename function to add_axioms_for_string_of_int

This is more uniform with the other builtin string functions

romainbrenguier committed:
Add builtin support for string_of_int

romainbrenguier committed:
Tests for String.valueOf(int)

smowton reviewed (commented):
Should include tests to check what happens when we *do* need the string representation? For example, `String.valueOf(100).charAt(1)` would presumably require us to make and then break a string-of-int object?

romainbrenguier committed:
Add version of make_string accepting iterators

This will make using make_string more flexible.

romainbrenguier commented:
@smowton I added a test for the case where the result matters in verification.

I also fixed the case where the radix is different from 10 (we can't use an existing c++ function for that case).

peterschrammel commented on src/solvers/refinement/string_builtin_function.cpp:
```diff
@@ -222,6 +222,85 @@ optionalt<exprt> string_insertion_builtin_functiont::eval(
+
+  exprt size_expr = from_integer(1, type);
+  std::size_t current_size = 2;
+  for(exprt min_int_with_current_size = radix; current_size <= upper_bound + 1;
```
To make this loop more readable I would only put `current_size` into`for()`.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 8de07de91854f7f0ba0640e6aa1fda627e1eee81).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80729163

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 3a1593b1ae07cdbcf24a29085c8a6a00c4363f08).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80735776
